kratos/utilities/exact_mortar_segmentation_utility.cpp

```diff
@@ -0,0 +1,1226 @@
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template< unsigned int TDim, unsigned int TNumNodes, bool TBelong>
+inline void ExactMortarIntegrationUtility<TDim, TNumNodes, TBelong>::ComputeClippingIntersections(
+    PointListType& PointList,
+    GeometryPointType& Geometry1,
+    GeometryPointType& Geometry2,
+    const PointType& RefCenter
+    )
+{
+    // We consider the Z coordinate constant
+    const double z_ref = RefCenter.Coordinate(3);
+    
+    // We find the intersection in each side
+    for (unsigned int i_edge = 0; i_edge < TNumNodes; i_edge++)
+    {  
+        const unsigned int ip_edge = (i_edge == (TNumNodes - 1)) ? 0 : i_edge + 1;
+        for (unsigned int j_edge = 0; j_edge < TNumNodes; j_edge++)
+        {
+            const unsigned int jp_edge = (j_edge == (TNumNodes - 1)) ? 0 : j_edge + 1;
+            
+            PointType intersected_point;
+            const bool intersected = Clipping2D(
+                intersected_point,
+                Geometry1[i_edge],
+                Geometry1[ip_edge],
+                Geometry2[j_edge],
+                Geometry2[jp_edge]
+                );
+            
+            if (intersected == true)
+            {
+                // Set the coordinate
+                intersected_point.Coordinate(3) = z_ref;
+                
+                // Ititialize the check
+                bool add_point = true;
+                for (unsigned int iter = 0; iter < PointList.size(); iter++)
+                {
+                    if (CheckPoints(intersected_point, PointList[iter]) == true)
+                    {
+                        add_point = false;
+                        break;
+                    }
+                }
+                
+                if (add_point == true) 
+                {
+                    if (TBelong == true) // NOTE: We do some kind of strange hash to know the intersected edges
+                    {
+                        const unsigned int hash = 2 * TNumNodes + 10 * i_edge + 100 * ip_edge + 1000 * j_edge + 10000 * jp_edge;
+                        PointList.push_back(PointBelong<TNumNodes>(intersected_point.Coordinates(), static_cast<BelongType>(hash))); 
+                    }
+                    else PointList.push_back(intersected_point);
+                }
+            }
+        }
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template< unsigned int TDim, unsigned int TNumNodes, bool TBelong>
+template<class TGeometryType>
+inline bool ExactMortarIntegrationUtility<TDim, TNumNodes, TBelong>::TriangleIntersections(
+    ConditionArrayListType& ConditionsPointsSlave,
+    PointListType& PointList,
+    TGeometryType& OriginalSlaveGeometry,
```

Comment:
some of those things can be const, i am sure of it

Reply:
Nope, IsInside is evil